query-translator/src/translate.rs

```diff
@@ -69,40 +70,45 @@ impl ToConstraint for ColumnConstraint {
         match self {
             EqualsEntity(qa, entid) =>
                 Constraint::equal(qa.to_column(), ColumnOrExpression::Entid(entid)),
+
             EqualsValue(qa, tv) =>
                 Constraint::equal(qa.to_column(), ColumnOrExpression::Value(tv)),
+
             EqualsColumn(left, right) =>
                 Constraint::equal(left.to_column(), right.to_column()),
+
             EqualsPrimitiveLong(table, value) => {
                 let value_column = QualifiedAlias(table.clone(), DatomsColumn::Value).to_column();
                 let tag_column = QualifiedAlias(table, DatomsColumn::ValueTypeTag).to_column();
-                Constraint::And {
-                    constraints: vec![
-                        Constraint::equal(value_column, ColumnOrExpression::Value(TypedValue::Long(value))),
-                        Constraint::In {
-                            left: tag_column,
-                            list:
-                                if value < 0 {
-                                    // Can't be a ref or a boolean.
-                                    vec![ColumnOrExpression::Integer(4),
-                                         ColumnOrExpression::Integer(5),]
-                                } else if value > 1 {
-                                    // Can't be a boolean.
-                                    vec![ColumnOrExpression::Integer(0),
-                                         ColumnOrExpression::Integer(4),
-                                         ColumnOrExpression::Integer(5),]
-                                } else {
-                                    vec![ColumnOrExpression::Integer(0),
-                                         ColumnOrExpression::Integer(1),
-                                         ColumnOrExpression::Integer(4),
-                                         ColumnOrExpression::Integer(5),]
-                                },
-                        },
-                    ],
+
+                /// A bare long in a query might match a ref, an instant, a long (obviously), or a
+                /// double. If it's negative, it can't match a ref, but that's OK -- it won't!
+                ///
+                /// However, '1' and '0' are used to represent booleans, and some integers are also
+                /// used to represent FTS values. We don't want to accidentally match those.
+                ///
+                /// We ask `SQLValueType` whether this value is in range for how booleans are
+                /// represented in the database.
+                ///
+                /// We only hit this code path when the attribute is unknown, so we're querying
+                /// `all_datoms`. That means we don't see FTS IDs at all -- they're transparently
+                /// replaced by their strings. If that changes, then you should also exclude the
+                /// string type code (10) here.
```

Comment:
It's not free to add new constraints to the query. I opted to file #363 instead — if we have a test that fails when we violate this assumption, then we should be good to go.